File: openFaces/source/org/openfaces/component/validation/FloatingIconMessage.java
/*
 * OpenFaces - JSF Component Library 3.0
 * Copyright (C) 2007-2012, TeamDev Ltd.
 * licensing@openfaces.org
 * Unless agreed in writing the contents of this file are subject to
 * the GNU Lesser General Public License Version 2.1 (the "LGPL" License).
 * This library is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * Please visit http://openfaces.org/licensing/ for more details.
 */
package org.openfaces.component.validation;

import org.openfaces.util.Resources;
import org.openfaces.util.ValueBindings;

import javax.faces.context.FacesContext;

/**
 * @author Vladimir Korenev
 */
public class FloatingIconMessage extends OUIMessage {
    public static final String COMPONENT_TYPE = "org.openfaces.FloatingIconMessage";

    private static final String DEFAULT_RENDERER_TYPE = "org.openfaces.FloatingIconMessage";
    private static final int DEFAULT_OFFSET_TOP = -4;
    private static final int DEFAULT_OFFSET_LEFT = -4;

    private String imageUrl;
    private Integer offsetTop;
    private Integer offsetLeft;
    private Boolean noImage;
    private Boolean noStyle;
    private String clientIdFor;

    private boolean runtimeDefined = false;

    public FloatingIconMessage() {
        setRendererType(DEFAULT_RENDERER_TYPE);
        setShowSummary(false);
    }

    public FloatingIconMessage(FloatingIconMessage template, boolean runtime) {
        super(template);
        runtimeDefined = runtime;
        setRendererType(DEFAULT_RENDERER_TYPE);
        if (template == null)
            return;
        if (template.getImageUrl() != null)
            imageUrl = template.getImageUrl();
        if (template.getOffsetLeft() != DEFAULT_OFFSET_LEFT)
            offsetLeft = template.getOffsetLeft();
        if (template.getOffsetTop() != DEFAULT_OFFSET_TOP)
            offsetTop = template.getOffsetTop();
        copyAttributes(template, "noImage", "noStyle", "showSummary", "showDetail");
    }

    public boolean isRuntimeDefined() {
        return runtimeDefined;
    }

    public String getClientIdFor() {
        return clientIdFor;
    }

    public void setClientIdFor(String clientIdFor) {
        this.clientIdFor = clientIdFor;
    }

    public boolean isNoImage() {
        return ValueBindings.get(this, "noImage", noImage, false);
    }

    public boolean isNoStyle() {
        return ValueBindings.get(this, "noStyle", noStyle, false);
    }

    public void setNoImage(boolean noImage) {
        this.noImage = noImage;
    }

    public void setNoStyle(boolean noStyle) {
        this.noStyle = noStyle;
    }

    public String getImageUrl() {
        return ValueBindings.get(this, "imageUrl", imageUrl,
                Resources.internalURL(getFacesContext(), "validation/error_icon.gif"));
    }

    public void setImageUrl(String imageUrl) {
        this.imageUrl = imageUrl;
    }

    public int getOffsetTop() {
        return ValueBindings.get(this, "offsetTop", offsetTop, DEFAULT_OFFSET_TOP);
    }

    public void setOffsetTop(int offsetTop) {
        this.offsetTop = offsetTop;
    }

    public int getOffsetLeft() {
        return ValueBindings.get(this, "offsetLeft", offsetLeft, DEFAULT_OFFSET_LEFT);
    }

    public void setOffsetLeft(int offsetLeft) {
        this.offsetLeft = offsetLeft;
    }

    @Override
    public Object saveState(FacesContext context) {
        return new Object[]{
                super.saveState(context),
                imageUrl,
                offsetTop,
                offsetLeft,
                noImage,
                noStyle
        };
    }

    @Override
    public void restoreState(FacesContext context, Object state) {
        Object values[] = (Object[]) state;
        int i = 0;
        super.restoreState(context, values[i++]);
        imageUrl = (String) values[i++];
        offsetTop = (Integer) values[i++];
        offsetLeft = (Integer) values[i++];
        noImage = (Boolean) values[i++];
        noStyle = (Boolean) values[i++];
    }


}

File: openFaces/source/META-INF/resources/openfaces/validation/floatingIconMessage.js
/*
 * OpenFaces - JSF Component Library 3.0
 * Copyright (C) 2007-2012, TeamDev Ltd.
 * licensing@openfaces.org
 * Unless agreed in writing the contents of this file are subject to
 * the GNU Lesser General Public License Version 2.1 (the "LGPL" License).
 * This library is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * Please visit http://openfaces.org/licensing/ for more details.
 */
O$._FloatingIconMessageRenderer = function(clientId, forClientId, imageUrl, topOffset, leftOffset, css, events,
                                           hideImg, showSummary, showDetail, defaultPresentation) {
  this.clientId = clientId;
  var messageElement = O$(this.clientId);
  if (messageElement) {
    var parentNodeElement = messageElement.parentNode;
    parentNodeElement.removeChild(messageElement);
  }
  this.forId = forClientId;
  this.imageUrl = imageUrl;
  this.topOffset = topOffset;
  this.leftOffset = leftOffset;
  this.errorCss = css;
  this.iconEvents = events;
  this.showImg = !hideImg;
  this.showSummary = showSummary;
  this.showDetail = showDetail;
  this.defaultPresentation = defaultPresentation;
};

O$.extend(O$._FloatingIconMessageRenderer.prototype, {
  renderMessage: function(message) {
    if (this.defaultPresentation && O$._isComponentHasPresentation(this.forId)) {
      return;
    }
    var forElement = O$.byIdOrName(this.forId);
    var messageElement = O$(this.clientId);
    if (message) {
      var messageText = "";
      if (this.showSummary) messageText = message.summary;
      if (this.showDetail) messageText = message.detail;
      if (this.showDetail && this.showSummary) messageText = message.summary + "; " + message.detail;
      if (!this.showDetail && !this.showSummary) messageText = message.summary;

      if (this.showImg) {
        if (!messageElement) {
          messageElement = document.createElement("img");
          messageElement.id = this.clientId;
          messageElement.src = this.imageUrl;
          messageElement.style.position = "absolute";
          O$.assignEvents(messageElement, this.iconEvents);
          document.body.appendChild(messageElement);
          O$.getClientMessageRenderersWithVisibleBubble()[this.clientId + "_" + this.forId] = this;
          O$.correctElementZIndex(messageElement, forElement, 0); // same as for the validated component
        }
        this.position();
        messageElement.alt = messageText;
        messageElement.title = messageText;

      }
      if (!forElement._of_titleChanged) {
        forElement._of_title = forElement.title;
        forElement._of_titleChanged = true;
      }

      forElement.title = messageText;

      if (this.errorCss) {
        O$.setStyleMappings(forElement, {validation_floatingIconMessage: this.errorCss});
        O$.repaintAreaForOpera(forElement, true);
      }

    } else {
      if (messageElement) {
        var parentNode = messageElement.parentNode;
        parentNode.removeChild(messageElement);
      }
      if (forElement._of_titleChanged) {
        forElement.title = forElement._of_title;
        forElement._of_title = undefined;
        forElement._of_titleChanged = undefined;
      }
      O$.setStyleMappings(forElement, {validation_floatingIconMessage: null});
      O$.repaintAreaForOpera(forElement, true);
    }
  },

  update: function() {
    var forElement = O$.byIdOrName(this.forId);
    var message;
    if (forElement) {
      message = O$.getMessages(forElement)[0];
      this.renderMessage(message);
    }
  },

  position: function() {
    var messageElement = O$(this.clientId);
    if (messageElement) {
      var forElement = O$.byIdOrName(this.forId);

      if (!O$.isVisibleRecursive(forElement)) {
        messageElement.style.display = "none";
        return;
      }
      messageElement.style.display = "";

      var elementPos = O$.getElementPos(forElement);
      var left = elementPos.x + this.leftOffset;
      var top = elementPos.y + this.topOffset;

      messageElement.style.left = left + "px";
      messageElement.style.top = top + "px";
    }
  }

});

